src/libmongoc/doc/mongoc_oidc_callback_params_t.rst

```diff
@@ -0,0 +1,215 @@
+    example_callback_fn (mongoc_oidc_callback_params_t *params)
+    {
+       user_data_t *user_data = (user_data_t *) mongoc_oidc_callback_params_get_user_data (params);
+
+       user_data->counter += 1;
+
+       // ...
+
+       if (/* ... */) {
+          user_data->error_message = "OIDC callback failed due to ...";
+          return NULL;
+       }
+
+       // ...
+    }
+
+    void
+    example (void)
+    {
+       mongoc_client_t *client = /* ... */;
+       bson_error_t error;
+
+       {
+          mongoc_oidc_callback_t *callback = mongoc_oidc_callback_new ();
+          mongoc_oidc_callback_set_fn (callback, &example_callback_fn);
+
+          user_data_t *user_data = malloc (sizeof (*user_data));
+          *user_data = (user_data_t){.counter = 0, .error_message = NULL};
+          mongoc_oidc_callback_set_user_data (callback, (void *) user_data);
+
+          mongoc_client_set_oidc_callback (client, callback);
+          mongoc_oidc_callback_destroy (callback);
+       }
+
+       // ... client operations ...
+
+       {
+          mongoc_oidc_callback_t *callback = mongoc_client_get_oidc_callback (client);
+          user_data_t *user_data = (user_data_t *) mongoc_oidc_callback_get_user_data (callback);
+
+          if (error.code != 0) {
+             printf ("client error message: %s\n", error.message);
+          }
+
+          if (user_data->error_message) {
+             printf ("custom error message: %s\n", user_data->error_message);
+          }
+
+          printf ("The callback function was invoked %n times!", user_data->counter);
+
+          free (user_data);
+       }
+
+       mongoc_client_destroy (client);
+    }
+
+Timeout
+```````
+
+The ``timeout`` parameter is used to determine when the callback function should report cancellation due to a timeout.
+
+When :symbol:`bson_get_monotonic_time()` is greater than ``timeout``, the callback function must invoke :symbol:`mongoc_oidc_callback_params_cancel_with_timeout()` and return ``NULL``.
+
+Username
+````````
+
+The ``username`` parameter is the value of the username component of the URI of the associated :symbol:`mongoc_client_t` or :symbol:`mongoc_client_pool_t` object from which the callback function is invoked.
+
+Timeout Cancellation
+````````````````````
+
+The ``cancel_with_timeout`` out parameter indicates cancellation of the callback function due to a timeout instead of an error.
+
+.. important::
+
+    The callback function MUST return ``NULL``, otherwise the invocation will be interpreted as a success even when ``cancel_with_timeout`` is set.
+
+.. code-block:: c
+
+    mongoc_oidc_credential_t *
+    example_callback_fn (mongoc_oidc_callback_params_t *params) {
+       const int64_t timeout = mongoc_oidc_callback_params_get_timeout (params);
+
+       if (bson_get_monotonic_time () > timeout) {
+          return mongoc_oidc_callback_params_cancel_with_timeout (params);
+       }
+
+       // ...
+    }
+
+Error Cancellation
```

Comment:
Suggest avoiding "cancellation". I expect an error could indicate many things (failure to connect to OIDC server, error in credentials, etc.). I interpret "cancellation" to mean something external requested cancellation.